src/System.Data.DataSetExtensions/tests/Mono/DataRowComparerTest.cs

```diff
@@ -0,0 +1,279 @@
+            {
+                // "#A2"
+                Assert.Equal(typeof(RowNotInTableException), ex.GetType());
+                // "#A3"
+                Assert.Null(ex.InnerException);
+                // "#A4"
+                Assert.NotNull(ex.Message);
+            }
+
+            // right row deleted
+            try
+            {
+                c.Equals(r2, r1);
+                Assert.True(false, "#B1");
+            }
+            catch (RowNotInTableException ex)
+            {
+                // "#B2"
+                Assert.Equal(typeof(RowNotInTableException), ex.GetType());
+                // "#B3"
+                Assert.Null(ex.InnerException);
+                // "#B4"
+                Assert.NotNull(ex.Message);
+            }
+
+            r2.Delete();
+
+            // both rows deleted
+            try
+            {
+                c.Equals(r2, r1);
+                Assert.True(false, "#C1");
+            }
+            catch (RowNotInTableException ex)
+            {
+                // #C2
+                Assert.Equal(typeof(RowNotInTableException), ex.GetType());
+                // #C3
+                Assert.Null(ex.InnerException);
+                // #C4
+                Assert.NotNull(ex.Message);
+            }
+        }
+
+        [Fact]
+        public void GetHashCodeWithVersions()
+        {
+            DataSet ds = new DataSet();
+            DataTable dt = new DataTable("MyTable");
+            ds.Tables.Add(dt);
+            dt.Columns.Add("col1");
+            dt.Columns.Add("col2");
+            DataRow r1 = dt.Rows.Add(new object[] { "foo", "bar" });
+            DataRow r2 = dt.Rows.Add(new object[] { "foo", "bar" });
+            ds.AcceptChanges();
+            DataRowComparer<DataRow> c = DataRowComparer.Default;
+            Assert.True(c.GetHashCode(r1) == c.GetHashCode(r2), "#1");
+            /*
+			// LAMESPEC: .NET fails here
+			r2 ["col2"] = "baz";
+			r2.AcceptChanges ();
+			Assert.IsFalse (c.GetHashCode (r1) == c.GetHashCode (r2), "#2");
+			ds.AcceptChanges (); // now r2 original value is "baz"
+			r2 ["col2"] = "bar";
+			Assert.IsFalse (c.GetHashCode (r1) == c.GetHashCode (r2), "#3");
+			// LAMESPEC: .NET fails here
+			DataRow r3 = dt.Rows.Add (new object [] {"foo", "baz"});
+			Assert.IsFalse (c.GetHashCode (r1) == c.GetHashCode (r3), "#4");
+			*/
+        }
+
+        [Fact]
+        public void GetHashCode_Row_Null()
+        {
+            DataRowComparer<DataRow> c = DataRowComparer.Default;
+
+            try
+            {
+                c.GetHashCode(null);
+                Assert.True(false, "#1");
+            }
+            catch (ArgumentNullException ex)
+            {
+                // #2
+                Assert.Equal(typeof(ArgumentNullException), ex.GetType());
+                // #3
+                Assert.Null(ex.InnerException);
+                // #4
+                Assert.NotNull(ex.Message);
+                // #5
```

Comment:
Didn't want to change much from the Mono source, presumably this was in their sources?